File: GFG/JavaScript/DSA/Easy/find-equal-point-in-string-of-brackets.js
// https://www.geeksforgeeks.org/problems/find-equal-point-in-string-of-brackets2542/1

class Solution {
  //Your code goes here
  findIndex(str) {
    let n = str.length;

    // Step 1: Calculate total number of closing brackets
    let totalClosing = 0;
    for (let i = 0; i < n; i++) {
      if (str[i] === ")") totalClosing++;
    }

    // Initialize counters
    let openingCount = 0;
    let closingCount = totalClosing;

    // Step 2: Iterate through the string
    for (let i = 0; i < n; i++) {
      if (str[i] === "(") {
        openingCount++;
      } else if (str[i] === ")") {
        closingCount--;
      }

      // Step 3: Check for the equal point
      if (openingCount === closingCount) {
        return i + 1; // Return 1-based index
      }
    }

    return 0; // No equal point found
  }
}

File: GFG/Python/DSA/Easy/find-equal-point-in-string-of-brackets.py
# https://www.geeksforgeeks.org/problems/find-equal-point-in-string-of-brackets2542/1

class Solution:
    def find_index(self, string):
        n = len(string)

        closing_brackets = 0
        for i in string:
            if i == ')':
                closing_brackets += 1

        opening_bracket = 0
        for i in range(0, n):
            if string[i] == '(':
                opening_bracket += 1
            else:
                closing_brackets -= 1

            if opening_bracket == closing_brackets:
                return i + 1
        return 0
